examples/jsm/renderers/nodes/core/NodeBuilder.js

```diff
@@ -0,0 +1,320 @@
+
+		let nodeUniform = nodeData.uniform;
+
+		if ( nodeUniform === undefined ) {
+
+			const uniforms = this.uniforms[ shaderStage ];
+			const index = uniforms.length;
+
+			nodeUniform = new NodeUniform( 'nodeU' + index, type, node );
+
+			uniforms.push( nodeUniform );
+
+			nodeData.uniform = nodeUniform;
+
+		}
+
+		return nodeUniform;
+
+	}
+
+	/*
+	analyzeNode( node ) {
+
+
+	}
+	*/
+
+	flowNode( node, output ) {
+
+		let flowData = {};
+		flowData.result = node.build( this, output );
+
+		return flowData;
+
+	}
+
+	_buildDefines( shader ) {
+
+		const defines = this.defines[ shader ];
+
+		let code = '';
+
+		for ( let name in defines ) {
+
+			code += `#define ${name} ${defines[ name ]}\n`;
+
+		}
+
+		return code;
+
+	}
+
+	getAttributesBodySnippet( /*shaderStage*/ ) {
+
+
+
+	}
+
+	getAttributesHeaderSnippet( /*shaderStage*/ ) {
+
+
+
+	}
+
+	getUniformsHeaderSnippet( shaderStage ) {
+
+		const uniforms = this.uniforms[ shaderStage ];
+
+		let snippet = '';
+
+		for ( let uniform of uniforms ) {
+
+			snippet += `${uniform.type} ${uniform.name}; `;
+
+		}
+
+		return snippet;
+
+	}
+
+	format( snippet, fromType, toType ) {
+
+		fromType = this.getVectorType( fromType );
+		toType = this.getVectorType( toType );
+
+		const typeToType = `${fromType} to ${toType}`;
+
+		switch ( typeToType ) {
+
+			case 'float to vec2' : return `vec2( ${snippet} )`;
+			case 'float to vec3' : return `vec3( ${snippet} )`;
+			case 'float to vec4' : return `vec4( vec3( ${snippet} ), 1.0 )`;
+
+			case 'vec2 to float' : return `${snippet}.x`;
+			case 'vec2 to vec3' : return `vec3( ${snippet}.x, ${snippet}.y, 0.0 )`;
+			case 'vec2 to vec4' : return `vec4( ${snippet}.x, ${snippet}.y, 0.0, 1.0 )`;
+
+			case 'vec3 to float' : return `${snippet}.x`;
+			case 'vec3 to vec2' : return `${snippet}.xy`;
+			case 'vec3 to vec4' : return `vec4( ${snippet}.x, ${snippet}.y, ${snippet}.z, 1.0 )`;
+
+			case 'vec4 to float' : return `${snippet}.x`;
+			case 'vec4 to vec2' : return `${snippet}.xy`;
+			case 'vec4 to vec3' : return `${snippet}.xyz`;
+
+		}
+
+		return snippet;
+
+	}
+
+	getHash() {
+
+		return this.vertexShader + this.fragmentShader;
+
+	}
+
+	build() {
+
+		const shaderStages = [ 'vertex', 'fragment' ];
+		const shaderData = {};
+
+		for ( let shaderStage of shaderStages ) {
+
+			this.shaderStage = shaderStage;
+
+			let slots = this.slots[ shaderStage ];
+
+			for ( let slot of slots ) {
+
+				let flowData = this.flowNode( slot.node, slot.output );
+
+				this.define( shaderStage, `NODE_${slot.name}`, flowData.result );
+
+			}
+
+		}
+
+		this.shaderStage = null;
+
+		for ( let shaderStage of shaderStages ) {
+
+			this.define( shaderStage, 'NODE_HEADER_UNIFORMS', this.getUniformsHeaderSnippet( shaderStage ) );
+			this.define( shaderStage, 'NODE_HEADER_ATTRIBUTES', this.getAttributesHeaderSnippet( shaderStage ) );
+			this.define( shaderStage, 'NODE_BODY_ATTRIBUTES', this.getAttributesBodySnippet( shaderStage ) );
```

Comment:
When looking at the code, does the distinction of header and body mean?

`NODE_HEADER_ATTRIBUTES`-> attribute declaration/definition
`NODE_BODY_ATTRIBUTES` -> attribute usage

Reply:
Yes, that's right.

A vertex shader example:
```
#version 450

layout(location = 0) in vec3 position;

NODE_HEADER_ATTRIBUTES
NODE_HEADER_UNIFORMS

void main(){
	NODE_BODY_ATTRIBUTES
	gl_Position = cameraUniforms.projectionMatrix * modelUniforms.modelViewMatrix * vec4( position, 1.0 );
}
```